lib/gosu.rb

```diff
@@ -17,10 +17,369 @@
+                                    :double, :double, :double, :double, :double, :double, :double, :double,
+                                    :_callback], :void
+  attach_function :Gosu_translate, [:double, :double, :_callback], :void
+  attach_function :Gosu_rotate,    [:double, :double, :double, :_callback], :void
+  attach_function :Gosu_scale,     [:double, :double, :double, :double, :_callback], :void
+  attach_function :Gosu_clip_to,   [:double, :double, :double, :double, :_callback], :void
+
+  attach_function :Gosu_gl_z,  [:double, :_callback],             :void
+  attach_function :Gosu_gl,    [:_callback],                      :void
+  attach_function :Gosu_render, [:int, :int, :_callback, :uint32], :pointer
+  attach_function :Gosu_record, [:int, :int, :_callback],          :pointer
+
+  attach_function :Gosu_button_down,       [:uint32], :bool
+  attach_function :Gosu_axis,              [:uint32], :double
+  attach_function :Gosu_button_id_to_char, [:uint32], :string
+  attach_function :Gosu_button_char_to_id, [:string], :uint32
+  attach_function :Gosu_button_name,       [:uint32], :string
+  attach_function :Gosu_gamepad_name,      [:uint32], :string
+
+  attach_function :Gosu_clipboard,      [], :string
+  attach_function :Gosu_set_clipboard,  [:string], :void
+
+  attach_function :Gosu_draw_line,     [:double, :double, :uint32, :double, :double, :uint32, :double, :uint32], :void
+  attach_function :Gosu_draw_triangle, [:double, :double, :uint32, :double, :double, :uint32,
+                                        :double, :double, :uint32, :double, :uint32],                            :void
+  attach_function :Gosu_draw_quad,     [:double, :double, :uint32, :double, :double, :uint32,
+                                        :double, :double, :uint32, :double, :double, :uint32, :double, :uint32], :void
+  attach_function :Gosu_draw_rect,     [:double, :double, :double, :double, :uint32, :double, :uint32],          :void
+
+  attach_function :Gosu_offset_x,   [:double, :double],                   :double
+  attach_function :Gosu_offset_y,   [:double, :double],                   :double
+  attach_function :Gosu_distance,   [:double, :double, :double, :double], :double
+  attach_function :Gosu_angle,      [:double, :double, :double, :double], :double
+  attach_function :Gosu_angle_diff, [:double, :double],                   :double
+  attach_function :Gosu_random,     [:double, :double],                   :double
+
+  attach_function :Gosu_screen_width,     [:pointer], :uint32
+  attach_function :Gosu_screen_height,    [:pointer], :uint32
+  attach_function :Gosu_available_width,  [:pointer], :uint32
+  attach_function :Gosu_available_height, [:pointer], :uint32
+
+  attach_function :Gosu_last_error, [], :string
+
+  def self.fps
+    Gosu_fps().tap { Gosu.check_last_error }
+  end
+
+  def self.flush
+    Gosu_flush()
+    Gosu.check_last_error
+  end
+
+  def self.milliseconds
+    Gosu_milliseconds().tap { Gosu.check_last_error }
```

Comment:
Using #tap is very time consuming because a Proc is created. You should avoid this…

Reply:
Maybe `Gosu.check_last_error` accepts optional a parameter which will be returned when no error was found. This seems to help and clarify your code in most cases…

Reply:
Thanks. I didn't know about the Proc creation, but I also had a feeling that this is doing too much on the Ruby side.

Reply:
In this very specific method it's probably better to skip the error handling altogether, and document it as being noexcept.

Reply:
I've changed check_last_error as you suggested. I left the error checking even in methods that cannot possibly throw because I was afraid it would be confusing to document which methods are noexcept, and which ones aren't.

However, I have removed the error checks from all `release` methods because C++ destructors really must never throw, and that's a rule that should be easy to follow in the wrapper code.